### PR TITLE
shared docker/util changes for launch integrations

### PR DESCRIPTION
**Pull request by stephchen**
Fixes WB-NNNN
Fixes #NNNN

Description
-----------
- pass resource in during launch spec creation instead of always defaulting to local
- have each runner construct its own docker image URI to avoid passing in platform-specific parameters (eg GCP docker URI requires GCP project name, GCP registry address etc and I'd rather not have to pass in too much of this sort of stuff)
- move env vars back into dockerfile instead of docker run command

Testing
-------
How was this PR tested?

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/launch/docker.py [hunk 1]
```diff
@@ -185,6 +197,30 @@ def build_docker_image_if_needed(
...
+        f"ENV WANDB_LAUNCH={True}",
+        f"ENV WANDB_LAUNCH_CONFIG_PATH={os.path.join(workdir,_project_spec.DEFAULT_LAUNCH_METADATA_PATH)}",
+        f"ENV WANDB_RUN_ID={launch_project.run_id or None}",
+        f"ENV WANDB_DOCKER={launch_project.docker_image}",
+    ])
+    dockerfile_contents += env_vars + "\n"
+
+    if runner_type == 'gcp-vertex':
+        # supply the entrypoint via the dockerfile since we don't `docker run`
+        dockerfile_contents += 'ENTRYPOINT ["python", "{entrypoint}"]\n'.format(entrypoint=launch_project.get_single_entry_point().name)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/docker.py:222]: This needs to have the override args added to it. See our function `get_entry_point_command`

Comment (stephchen) [reply wandb/sdk/launch/docker.py:222]: gcp takes overrides as a separate arg at the job construction level so i thought i'd keep it there

File: wandb/sdk/launch/runner/local.py [hunk 2]
```diff
@@ -108,15 +109,19 @@ def run(self, launch_project: LaunchProject) -> Optional[AbstractRun]:
...
 
         if launch_project.docker_image is None or launch_project.build_image:
+            image_uri = construct_local_image_uri(launch_project)
             image = build_docker_image_if_needed(
                 launch_project=launch_project,
                 api=self._api,
                 copy_code=copy_code,
                 workdir=container_workdir,
                 container_env=container_env,
+                runner_type='local',
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/local.py:119]: Confused by this, won't this make it so the check for `gcp_vertex` (L220) in `build_docker_image_if_needed` never be true?

Comment (stephchen) [reply wandb/sdk/launch/runner/local.py:119]: this is in the local runner, this PR doesn't have the vertex runner in it — looks like i left the gcp line in `docker.py` in by accident lol you're right that it won't get hit in any of this code, i can take it out so the PR is cleaner

Comment (KyleGoyette) [reply wandb/sdk/launch/runner/local.py:119]: Ah I see now.

**Review by KyleGoyette — APPROVED** [review]
LGTM nice work

**Pull request merged**